pkg/codegen/schema/bind.go

```diff
@@ -1049,6 +1049,37 @@ func bindDefaultValue(path string, value interface{}, spec *DefaultSpec, typ Typ
 	return dv, diags
 }
 
+func isOverlyNested(path string, typ Type) hcl.Diagnostics {
```

Comment:
This feels like a nit, but I don't understand why you unrolled the loop and incurred an additional function call (however trivial)

I'm not sure that there is a readability difference between your version and the below, but I tend to favor shorter code I guess

``` go
func isOverlyNested(path string, typ Type) hcl.Diagnostics {
        curr := typ
        depth := 0
        isInnerMostAny := false

        for ; curr != nil; depth += 1 {
                isInnerMostAny = curr == AnyType
                switch t := curr.(type) {
                case *ArrayType:
                        curr = t.ElementType
                case *MapType:
                        curr = t.ElementType
                default:
                        curr = nil             
                }
        }

        if depth <= 3 || (isInnerMostAny && depth <= 4) {
                return nil
        }
        return hcl.Diagnostics{errorf(path, "type %v is nested too deeply (%d)", typ, depth))
}
```

Reply:
> but I tend to favor shorter code I guess

Yeh generally so do I, but in this case I felt the explicit three steps made it easier to read than trying to reason what a loop was doing.

Reply:
Well turns out the check is more complex and a loop is actually needed, unbounded depth is fine for named types.